_posts/2021-01-08-consumer-tuning.md

```diff
@@ -0,0 +1,281 @@
+* A new consumer group is formed
+* The designated group leader assigns partitions evenly to group members
+* The consumers in the group receive their assignments and start to fetch data
+
+Obviously, the rebalancing process takes time. And during a rolling restart of a consumer group cluster, it happens repeatedly. So rebalancing can have a clear impact on the performance of your cluster group. As mentioned, increasing the number of heartbeat checks reduces the likelihood of unnecessary rebalances. But there a couple more approaches you can take. One is to set up _static membership_ to reduce the overall number of rebalances. Static membership uses persistence so that a consumer instance is recognized during a restart after a session timeout.
+
+You can use the `group.instance.id` property to specify a unique group instance id for a consumer. The consumer group coordinator can then use the id when identifying a new consumer instance following a restart.
+
+```properties
+# ...
+group.instance.id=UNIQUE-ID
+# ...
+```
+
+The consumer group coordinator assigns the consumer instance a new member id, but as a static member it continues with the same instance id, and receives the same assignment of topic partitions is made.
+
+You can also calibrate the `max.poll.interval.ms` and `max.poll.interval.ms` polling configuration properties to reduce the number of rebalances.
+
+`max.poll.interval.ms`
+: Sets the interval to check the consumer is continuing to process messages. If the consumer application does not make a call to poll at least every `max.poll.interval.ms` milliseconds, the consumer is considered to be failed, causing a rebalance. If the application cannot process all the records returned from poll in time, you can avoid a rebalance by using this property to increase the interval in milliseconds between polls for new messages from a consumer.
+
+`max.poll.records`
+: Sets the number of processed records returned from the consumer. You can use the `max.poll.records` property to set a maximum limit on the number of records returned from the consumer buffer, allowing your application to process fewer records within the `max.poll.interval.ms` limit.
+
+```properties
+# ...
+group.instance.id=UNIQUE-ID
+max.poll.interval.ms=300000
+max.poll.records=500
+# ...
+```
+
+### Changing your consumer habits
+
+So that was a brief run through some of the most frequently used configuration options. Any strategy you choose to adopt will always have to take into account the other _moving parts_ -- the intended functionality of your consumer client application, the broker architecture, and the producer configuration. For example, if you are not using transactional IDs with your producers, then you probably won't see a need for using the `isolation.level` property. But it is unlikely that there isn't anything you can do to make your consumers run better. And the consumer configuration options are available exactly for that reason.
```

Comment:
```suggestion
So that was a brief run through some of the most frequently used configuration options. Any strategy you choose to adopt will always have to take into account the other _moving parts_ -- the intended functionality of your consumer client application, the broker architecture, and the producer configuration. For example, if you are not using transactional producers, then there's no point in setting the `isolation.level` property. But it is unlikely that there's  nothing you can do to make your consumers run better. And the consumer configuration options are available exactly for that reason.
```